### PR TITLE
Implement Category, Topic, Subtopic, Feedback

### PR DESCRIPTION
**Pull request by maxxkut**

### CODE REVIEW

**Review by NeHo4y — APPROVED** [review]

File: server/src/main/kotlin/com/github/neho4y/category/domain/Topic.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.github.neho4y.category.domain
+
+import java.util.*
```
Comment (dxahtepb) [inline server/src/main/kotlin/com/github/neho4y/category/domain/Topic.kt]: You have unused imports here and in TopicRepository too

File: server/src/main/kotlin/com/github/neho4y/category/service/impl/SubtopicServiceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            throw RuntimeException();
+        }
+        subtopic.description = subtopicDto.description
+        subtopic.topicId = subtopicDto.topicId
+        subtopicRepository.save(subtopic)
+        log.info("Subtopic ${subtopic.id} is updated with description ${subtopicDto.description}")
+    }
+
+    override fun deleteSubtopic(id: Long): Subtopic {
+        val subtopic = subtopicRepository.findByIdAndIsDeletedFalse(id).orElseThrow();
```
Comment (dxahtepb) [inline server/src/main/kotlin/com/github/neho4y/category/service/impl/SubtopicServiceImpl.kt]: You forgot the semicolon here

File: server/src/main/kotlin/com/github/neho4y/category/service/impl/CategoryServiceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import com.github.neho4y.category.model.CategoryDto
+import com.github.neho4y.category.service.CategoryService
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import java.lang.RuntimeException
+
+
+@Service
+class CategoryServiceImpl(
+        private val categoryRepository: CategoryRepository) : CategoryService {
```
Comment (dxahtepb) [inline server/src/main/kotlin/com/github/neho4y/category/service/impl/CategoryServiceImpl.kt]: Check class constructor formatting here. If it is multiline, then closing bracket should be on a separate line.

File: server/src/main/kotlin/com/github/neho4y/feedback/domain/repository/FeedbackRepository.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.github.neho4y.feedback.domain.repository
+
+import com.github.neho4y.feedback.domain.Feedback
+import org.springframework.data.jpa.repository.JpaRepository
+
+interface FeedbackRepository : JpaRepository<Feedback, Long>{
```
Comment (dxahtepb) [inline server/src/main/kotlin/com/github/neho4y/feedback/domain/repository/FeedbackRepository.kt]: You don't need to write a body for an empty class

File: server/src/test/kotlin/com/github/neho4y/category/service/CategoryServiceIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @Test
+    fun `When add category with the same description then fail`() {
+
+        // given
+        val categoryCreationDto = CategoryCreationDto("Test category")
+        val sameCategory = categoryCreationDto.copy()
+        categoryService.createCategory(categoryCreationDto)
+
+        // when
+        val exception = assertThrows<Exception> {
```
Comment (dxahtepb) [inline server/src/test/kotlin/com/github/neho4y/category/service/CategoryServiceIntegrationTest.kt:27]: You don't use this variable at all

File: server/src/test/kotlin/com/github/neho4y/category/service/SubtopicServiceIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    private var topicId: Long = 0
+
+    @Autowired
+    private lateinit var categoryRepository: CategoryRepository
+    private var categoryId: Long = 0
+
+    @BeforeEach
+    fun generateTestData() {
+        val category = Category("Test category")
+        categoryRepository.save(category)
```
Comment (dxahtepb) [inline server/src/test/kotlin/com/github/neho4y/category/service/SubtopicServiceIntegrationTest.kt:34]: You may use `saveAndFlush()` here

**Review by dxahtepb — CHANGES_REQUESTED** [review]
Looks good! But please do some formatting fixes

**Comment by dxahtepb** [comment]
Looks like we forgot about methods in services to retrieve all Categories, Topics etc.

**Review by dxahtepb — CHANGES_REQUESTED** [review]
You forgot methods in services to retrieve all Topics, Categories, etc...

**Review by dxahtepb — APPROVED** [review]

**Review by NeHo4y — APPROVED** [review]

File: server/src/main/kotlin/com/github/neho4y/category/model/SubtopicDto.kt [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.github.neho4y.category.model
+
+data class SubtopicCreationDto(
+
```
Comment (Walingar) [inline server/src/main/kotlin/com/github/neho4y/category/model/SubtopicDto.kt:4]: Лишние отступы

File: server/src/main/kotlin/com/github/neho4y/category/service/impl/CategoryServiceImpl.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+@Service
+class CategoryServiceImpl(
+    private val categoryRepository: CategoryRepository
+) : CategoryService {
+
+    private val log = LoggerFactory.getLogger(this::class.java)
+
+    override fun createCategory(categoryCreationDto: CategoryCreationDto): Category {
+        if (categoryRepository.existsByDescriptionAndIsDeletedFalse(categoryCreationDto.description)) {
+            throw RuntimeException()
```
Comment (Walingar) [inline server/src/main/kotlin/com/github/neho4y/category/service/impl/CategoryServiceImpl.kt:22]: Более специфичное бросать что-то?

File: server/src/main/kotlin/com/github/neho4y/feedback/domain/repository/FeedbackRepository.kt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.github.neho4y.feedback.domain.repository
+
+import com.github.neho4y.feedback.domain.Feedback
+import org.springframework.data.jpa.repository.JpaRepository
+
+interface FeedbackRepository : JpaRepository<Feedback, Long> {
```
Comment (Walingar) [inline server/src/main/kotlin/com/github/neho4y/feedback/domain/repository/FeedbackRepository.kt:6]: {} не нужны

**Pull request merged**